File: frontend/src/services/chat.service.ts
import { api } from "@/lib/api-client";
import { ENDPOINTS } from "@/constants/api";
import { PageResponse } from "./post.service";
import { 
  ConversationResponse, 
  MessageResponse, 
  PresenceResponse, 
  PresignedUrlResponse 
} from "@/types";

export interface PresignedUrlRequest {
  fileName: string;
  fileSize: number;
  contentType: string;
}

export const chatService = {
  getConversations: async () => {
    const res = await api.get<ConversationResponse[]>(ENDPOINTS.CHAT.CONVERSATIONS);
    return res;
  },

  createConversation: async (memberIds: string[], type: "DIRECT" | "GROUP", name?: string) => {
    const res = await api.post<ConversationResponse>(ENDPOINTS.CHAT.CONVERSATIONS, {
      memberIds,
      type,
      name
    });
    return res;
  },

  getMessages: async (conversationId: string, page = 0, size = 30) => {
    const res = await api.get<PageResponse<MessageResponse>>(
      ENDPOINTS.CHAT.MESSAGES(conversationId),
      { params: { page, size } }
    );
    return res;
  },

  getPresignedUrl: async (payload: PresignedUrlRequest) => {
    const res = await api.post<PresignedUrlResponse>(
      ENDPOINTS.CHAT.ATTACHMENT_URL,
      payload
    );
    return res;
  },

  getUserPresence: async (userId: string) => {
    const res = await api.get<PresenceResponse>(
      ENDPOINTS.CHAT.PRESENCE(userId)
    );
    return res;
  },

  uploadAttachmentFile: async (
    uploadUrl: string,
    file: File,
    onProgress?: (progressPercent: number) => void
  ): Promise<void> => {
    return new Promise((resolve, reject) => {
      const xhr = new XMLHttpRequest();
      xhr.open("PUT", uploadUrl, true);
      xhr.setRequestHeader("Content-Type", file.type);

      if (xhr.upload && onProgress) {
        xhr.upload.onprogress = (event) => {
          if (event.lengthComputable) {
            const percentComplete = Math.round((event.loaded / event.total) * 100);
            onProgress(percentComplete);
          }
        };
      }

      xhr.onload = () => {
        if (xhr.status === 200) {
          resolve();
        } else {
          reject(new Error(`Upload failed with status: ${xhr.status}`));
        }
      };

      xhr.onerror = () => {
        reject(new Error("Network error during file upload to storage."));
      };

      xhr.send(file);
    });
  }
};

File: backend/src/main/java/com/vibecart/api/modules/chat/service/ChatService.java
package com.vibecart.api.modules.chat.service;

import com.vibecart.api.common.dto.PageResponse;
import com.vibecart.api.modules.chat.dto.request.ConversationRequest;
import com.vibecart.api.modules.chat.dto.request.MessageRequest;
import com.vibecart.api.modules.chat.dto.request.PresignedUrlRequest;
import com.vibecart.api.modules.chat.dto.request.TypingRequest;
import com.vibecart.api.modules.chat.dto.response.ConversationResponse;
import com.vibecart.api.modules.chat.dto.response.MessageResponse;
import com.vibecart.api.modules.chat.dto.response.PresignedUrlResponse;

import java.util.List;
public interface ChatService {
    ConversationResponse createOrGetConversation(ConversationRequest request, String currentUsername);
    List<ConversationResponse> getConversationsForUser(String currentUsername);
    PageResponse<MessageResponse> getMessages(String conversationId, int page, int size, String currentUsername);
    MessageResponse saveAndBroadcastMessage(MessageRequest request, String currentUsername);
    void broadcastTyping(TypingRequest request, String currentUsername);
    PresignedUrlResponse generateAttachmentUploadUrl(PresignedUrlRequest request, String currentUsername);
    void markConversationAsRead(String conversationId, String currentUsername);
}
